opentelemetry-sdk/src/logs/log_emitter.rs

```diff
@@ -136,19 +159,10 @@ impl LoggerProvider {
     }
 }
 
-#[derive(Debug)]
-struct LoggerProviderInner {
-    processors: Vec<Box<dyn LogProcessor>>,
-    resource: Resource,
-}
-
 impl Drop for LoggerProviderInner {
     fn drop(&mut self) {
-        for processor in &mut self.processors {
-            if let Err(err) = processor.shutdown() {
-                global::handle_error(err);
-            }
-        }
+        let _ = self.shutdown();
```

Comment:
this is still calling shutdown, even when is_shutdown is already true?

Reply:
shutdown take care of checking is_shutdown, so we dont need to do it here

Reply:
it should still be done in drop to avoid the false alarm to users, as shutdown() will now emit an internal log about "already shutdown" if user has already invoked shutdown.
See https://github.com/open-telemetry/opentelemetry-rust/pull/2156/files for reference.

Reply:
This already have a TODO in next line to be taken care in #2184

Reply:
This is a logical flaw that drop is unnecessarily calling shutdown, when it is not supposed to. 2184 just fixes the confusing internal log, but isn't the right behavior is for drop to *not* call shutdown itself?

Reply:
I am not sure why this is a flaw - The MeterProvider has a flaw for checking the is_shutdown flag twice - once is drop, and then in shutdown. I think we should fix that. The `Inner::shutdown` should check the is_shutdown, and return the Result status to the caller, caller then handles the Error accordingly.

Reply:
I have instead removed is_shutdown check from LoggerProviderInner::shutdown(), and added both in LoggerProviderInner::drop and LoggerProvider::Shutdown. Let me know if this works, will update TracerProvider PR too accordingly.

Reply:
I am not seeing any flaw with MeterProvider. It is expected that shutdown() checks if already shutdown and return appropriate Result. It is also expected that Drop should not even call shutdown() if user has already done so. In my view, calling shutdown from drop is SDK being good citizen. If user has forgotten to call shutdown, then drop will do it for them. But if user has already done their job, drop should detect that and don't call shutdown again.

If you can send a PR with a proposal to address issue with MeterProvider, that could help me see what is the current issue vs proposed solution.

Reply:
> If you can send a PR with a proposal to address issue with MeterProvider, that could help me see what is the current issue vs proposed solution.

Please review the updated changes in this PR. If it looks good, I will apply similar changes for TracerProvider and MeterProvider. The issue with MeterProvider is that the is_shutdown check is performed twice during a drop, which is unnecessary:
```
MeterProviderInner::Drop -> checks is_shutdown -> calls MeterProviderInner::shutdown -> checks is_shutdown again -> invokes shutdown

MeterProvider::shutdown -> calls MeterProviderInner::shutdown -> checks is_shutdown -> invokes shutdown
```

The `MeterProviderInner::shutdown` method is private and won’t be called directly by the user. We have two possible approaches for handling the is_shutdown check:

- Option 1: Perform the is_shutdown check within `MeterProviderInner::shutdown` and skip the check in both `MeterProviderInner::Drop` and` MeterProvider::shutdown`.
- Option 2: Skip the is_shutdown check in `MeterProviderInner::shutdown` and instead perform the check in both `MeterProviderInner::Drop` and `MeterProvider::shutdown`.

I earlier implemented option 1 in this PR, but now changed it to option 2 based on your concern.

Reply:
This looks good!